### PR TITLE
ENG-1074: add pagination to session log records

### PR DESCRIPTION
**Pull request by efiShtain**
-- added sequence to session log record
-- added custom migrations to update filled databases with correct seq field
-- updated the primary key on session log records to (session_id, seq)
-- added filter with sort order (asc/desc)
-- updated cli
-- updated tests

### CODE REVIEW

File: internal/backend/db/dbgorm/sessions.go [hunk 1]
```diff
@@ -93,18 +128,26 @@ func (db *gormdb) UpdateSessionState(ctx context.Context, sessionID sdktypes.Ses
...
 func addSessionLogRecordDB(tx *gorm.DB, logr *scheme.SessionLogRecord) error {
-	return tx.Create(logr).Error
+	// This is a hack in gorm so we can use subquery when creating an object
+	// We want to have a sequence number per session
+	// The insert uses a subquery to get the current max sequence for this session_id
+	// and insert the next value in one query
+	data := map[string]interface{}{}
+	data["SessionID"] = logr.SessionID
+	data["Data"] = logr.Data
+	data["Seq"] = clause.Expr{SQL: "(select COALESCE(MAX(seq), 0)+1 from session_log_records where session_id = ?)", Vars: []interface{}{logr.SessionID}} // fmt.Sprintf("select COALESCE(MAX(seq), 0) from session_log_records where session_id = '%s'", logr.SessionID)
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions.go]: fmt

File: internal/backend/db/dbgorm/sessions.go [hunk 2]
```diff
@@ -38,16 +38,51 @@ func (db *gormdb) DeleteSession(ctx context.Context, sessionID sdktypes.SessionI
...
 	}
 
 	prs, err := kittehs.TransformError(rs, scheme.ParseSessionLogRecord)
+	log := sdktypes.NewSessionLog(prs)
+
+	nextPageToken := ""
+	if len(rs) == int(filter.PageSize) && len(rs) > 0 {
+		nextPageToken = fmt.Sprintf("%d", rs[len(rs)-1].Seq)
+	}
+	fmt.Println("Next Page Token", nextPageToken)
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions.go]: fmt

File: proto/autokitteh/sessions/v1/svc.proto [hunk 3]
```diff
@@ -71,10 +71,23 @@ message GetResponse {
...
+  int32 page_size = 20;
+
+  int32 skip = 21 [(buf.validate.field).cel = {
+    id: "session.list.skip"
+    message: "Must be >= 0"
+    expression: "this >= 0"
+  }];
+
+  string page_token = 22;
+  bool ascending = 11;
```
Comment (ig-ra) [inline proto/autokitteh/sessions/v1/svc.proto:83]: is_accending? in_accending_order?

File: internal/backend/gormkitteh/z.go [hunk 4]
```diff
@@ -15,7 +15,7 @@ func OpenZ(z *zap.Logger, cfg *Config, f func(*gorm.Config)) (*gorm.DB, error) {
 	return Open(cfg, func(c *gorm.Config) {
 		l := &zapgorm2.Logger{
 			ZapLogger:                 z,
-			LogLevel:                  logger.Warn,
+			LogLevel:                  logger.Info,
```
Comment (ig-ra) [inline internal/backend/gormkitteh/z.go]: revert to warn?

Comment (efiShtain) [reply internal/backend/gormkitteh/z.go]: I'll revert, forgot it on for debug
good catch

File: internal/backend/db/dbgorm/sessions.go (hunk 1 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions.go]: removed

File: internal/backend/db/dbgorm/sessions.go (hunk 2 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions.go]: removed

File: proto/autokitteh/sessions/v1/svc.proto (hunk 3 above)
Comment (efiShtain) [reply proto/autokitteh/sessions/v1/svc.proto:83]: in idiomatic go you shouldn't add is prefix 
and the context is listing, so it is known to be order related

File: proto/autokitteh/sessions/v1/svc.proto [hunk 5]
```diff
@@ -71,10 +71,23 @@ message GetResponse {
 
 message GetLogRequest {
   string session_id = 1 [(buf.validate.field).string.min_len = 1];
+  int32 page_size = 20;
```
Comment (itayd) [inline proto/autokitteh/sessions/v1/svc.proto:74]: didn't we say we'll use PaginationOptions or some such?

Comment (efiShtain) [reply proto/autokitteh/sessions/v1/svc.proto:74]: Nope

We said we to design a generic thing and do it once all over

Currently there is no more requirement for pagination, only two 

Third request is the time to generalize

File: internal/backend/db/dbgorm/gorm_test.go [hunk 6]
```diff
@@ -137,9 +137,9 @@ func setupDB(config *gormkitteh.Config) *gorm.DB {
 	logger := logger.New(
 		log.New(os.Stdout, "\r\n", log.LstdFlags), // io writer
 		logger.Config{
-			SlowThreshold: time.Second,   // Slow SQL threshold
-			LogLevel:      logger.Silent, // Log level
-			Colorful:      false,         // Disable color
+			SlowThreshold: time.Second, // Slow SQL threshold
+			LogLevel:      logger.Info, // Log level
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/gorm_test.go]: I will revert this, ok?

Comment (efiShtain) [reply internal/backend/db/dbgorm/gorm_test.go]: Sure

File: proto/autokitteh/sessions/v1/svc.proto (hunk 3 above)
Comment (tebeka) [reply proto/autokitteh/sessions/v1/svc.proto:83]: Agree with @efiShtain about naming. You can also define a enum for ascending or descending and use it.

File: internal/backend/db/dbgorm/sessions.go [hunk 7]
```diff
@@ -165,10 +170,7 @@ func (gdb *gormdb) getSessionLogRecords(ctx context.Context, filter sdkservices.
 			}
 		}
 
-		if err := q.Model(&scheme.SessionLogRecord{}).Count(&n).Error; err != nil {
-			return err
-		}
-
+		// Default is descnding order
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions.go]: typo

File: internal/backend/db/dbgorm/sessions_test.go [hunk 8]
```diff
@@ -64,12 +65,18 @@ func testLastLogRecord(t *testing.T, f *dbFixture, numRecords int, sessionID sdk
...
+	assert.Equal(t, a, b)
+}
+
 func TestCreateSession(t *testing.T) {
 	f := preSessionTest(t)
 
 	// test createSession
 	s := f.newSession(sdktypes.SessionStateTypeCompleted)
-	f.createSessionsAndAssert(t, s) // all session assets are optional and will be set to nil
+	f.createSessionsAndAssert(t, s) // all session asserts are optional and will be set to nil
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions_test.go]: assets :) not asserts. e.g. build, env, deployment, event

File: internal/backend/db/dbgorm/sessions_test.go [hunk 9]
```diff
@@ -79,7 +86,7 @@ func TestCreateSessionForeignKeys(t *testing.T) {
 	// check session creation if foreign keys are not nil
 	f := preSessionTest(t)
 
-	// test with existing assets
+	// test with existing asserts
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions_test.go]: same. assets

File: internal/backend/db/dbgorm/sessions_test.go [hunk 10]
```diff
@@ -94,7 +101,7 @@ func TestCreateSessionForeignKeys(t *testing.T) {
 	f.createEventsAndAssert(t, evt)
 	f.createSessionsAndAssert(t, s)
 
-	// negative test with non-existing assets
+	// negative test with non-existing asserts
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions_test.go]: same. assets

File: internal/backend/db/dbgorm/sessions_test.go [hunk 11]
```diff
@@ -251,3 +258,165 @@ func TestAddSessionPrintLogRecord(t *testing.T) {
 
 	testLastLogRecord(t, f, 2, s.SessionID, l)
 }
+
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions_test.go:261]: plz combine two tests for order into one with parametrized. see events test for example

File: internal/backend/db/dbgorm/sessions_test.go [hunk 12]
```diff
@@ -251,3 +258,165 @@ func TestAddSessionPrintLogRecord(t *testing.T) {
...
+	assert.NoError(t, err)
+	assert.Equal(t, len(logs), 1)
+	assert.Equal(t, n, int64(2))
+}
+
+func TestSessionLogRecordSkipAll(t *testing.T) {
+	f := preSessionTest(t)
+
+	s := f.newSession(sdktypes.SessionStateTypeCompleted)
+	l := sdktypes.NewPrintSessionLogRecord("meow")
```
Comment (ig-ra) [inline internal/backend/db/dbgorm/sessions_test.go:330]: plz move common part to function

**Review by ig-ra — APPROVED** [review]

File: internal/backend/db/dbgorm/sessions.go (hunk 7 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions.go]: fixed

File: internal/backend/db/dbgorm/sessions_test.go (hunk 8 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions_test.go]: fixed

File: internal/backend/db/dbgorm/sessions_test.go (hunk 9 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions_test.go]: fixed

File: internal/backend/db/dbgorm/sessions_test.go (hunk 10 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions_test.go]: fixed

File: internal/backend/db/dbgorm/sessions_test.go (hunk 11 above)
Comment (efiShtain) [reply internal/backend/db/dbgorm/sessions_test.go:261]: fixed

File: go.sum [hunk 13]
```diff
@@ -58,6 +58,8 @@ github.com/PuerkitoBio/goquery v1.9.2 h1:4/wZksC3KgkQw7SQgkKotmKljk0M6V8TUvA8Wb4
 github.com/PuerkitoBio/goquery v1.9.2/go.mod h1:GHPCaP0ODyyxqcNoFGYlAprUFH81NuRPd0GX3Zu2Mvk=
 github.com/adrg/xdg v0.4.0 h1:RzRqFcjH4nE5C6oTAxhBtoE2IRyjBSa62SCbyPidvls=
 github.com/adrg/xdg v0.4.0/go.mod h1:N6ag73EX4wyxeaoeHctc1mas01KZgsj5tYiAIwqJE/E=
+github.com/alecthomas/kong v0.7.1 h1:azoTh0IOfwlAX3qN9sHWTxACE2oV8Bg2gAwBsMwDQY4=
```
Comment (daabr) [inline go.sum:61]: No changes in `go.mod`?

File: internal/backend/sessions/sessions.go [hunk 14]
```diff
@@ -54,8 +54,12 @@ func (s *sessions) StartWorkers(ctx context.Context) error {
 	return nil
 }
 
-func (s *sessions) GetLog(ctx context.Context, sessionID sdktypes.SessionID) (sdktypes.SessionLog, error) {
-	return s.svcs.DB.GetSessionLog(ctx, sessionID)
+// func (s *sessions) GetLog(ctx context.Context, sessionID sdktypes.SessionID) (sdktypes.SessionLog, error) {
```
Comment (daabr) [inline internal/backend/sessions/sessions.go]: Why is this commented out?

File: internal/backend/sessionsgrpcsvc/svc.go [hunk 15]
```diff
@@ -102,15 +102,25 @@ func (s *server) GetLog(ctx context.Context, req *connect.Request[sessionsv1.Get
 		return nil, sdkerrors.AsConnectError(err)
 	}
 
-	hist, err := s.sessions.GetLog(ctx, sessionID)
+	filter := sdkservices.ListSessionLogRecordsFilter{
```
Comment (daabr) [inline internal/backend/sessionsgrpcsvc/svc.go:105]: Shouldn't the implementation of paging here in `GetLog` be the same as in `List`? Both in terms in input params, and backend handling of them.

BTW, regarding the existing pagination in `List` (not necessarily related to this PR):  if the server can quietly override the user's page size (which is good and important): shouldn't the response also include the actual page size?

File: go.sum (hunk 13 above)
Comment (efiShtain) [reply go.sum:61]: I don't remember changing anything

File: internal/backend/sessions/sessions.go (hunk 14 above)
Comment (efiShtain) [reply internal/backend/sessions/sessions.go]: left overs
removed
thanks

**Pull request merged**